### DISCUSSION TITLE
CustomAttributeType reduces union type to primitive type

### DISCUSSION THREAD

**Discussion by zirkelc** [tywalch/electrodb#313] [Q&A] [answered]
When I give the `CustomAttributeType()` a union type like this:

```ts
type PersonnelRole = 
| string
| {
    type: 'contractor';
    contractStartDate: string;
    contractEndDate: string;
};
```

The resulting type on the entity is reduces to the primitive type `string`. See this [ElectroDB example](https://electrodb.fun/?#code/JYWwDg9gTgLgBAbzgUQHY2DAngGjgYQFcBnGCEAQRhimACNCYBTAFSzCbgF84AzKcnADkTADZMAxjQgATOkIDcAKCUSIqUnACqqYAEdCTANJMsAZSwg6EUQC44hXQc7FL10XAC8cC1ZsAKAEplbA4UcFEILCYmAEkAES84UlpUAHM4ADJEAG0dfUMTczcbAF17AENULC5lVXVNfOcKNKZfd3tHApcSj292gOClUM4AWXUYAAsk1EIrJigs3KbDFrbe8rgqmrqRuAAFBeJ1VDEAJRtObyUAH2SaYHTbxCU4N7gR+yE1dCgKqWgile7x+NH+MDMMAqsHiFWY9hSjzSynecFBfykaBksPh91SyKUtRUP00ULo4iSQgA7tAANZgUT-JgAfTJ4iB9Q08A4UGOqBmTCpKHQmCw-gQwLeIFkYnsEtRqKYIuwXx5fNOoiEOElCuICwAbsAJEwvjSoPTGcatTrUfqjsB1F8AIxCHVcbWouEPBjMYhym3vJgRKIxWIyf0KyMfdgmggkMiUai0H2sGMAHmQweicXiAD5-EJEekhIEA293WW4EGGSGFhGo6jPsIi2lXQ2Kw3eMBeTAAHIVECx+UN95NwsPYsekdvKBMAzdpjhj5QQxTyMdqOM0j9wf16djlvW6czueEBdLmirysbyMVVp7kdj2bzKBtqM3hV3uKocboSZ+l5jzgKk4QkSZ7ByIQvyEUo10fGN7CIUhyCob1GFTDg01-KZ8yEZ86AWEs4IbPUYHsIIvFzRAPwbGQmF4CpCFEMi4AozwqIQGio1aFj-GZexsMmPAEC-Lg5S-ex8IWLhAkowCgPeWcYEIKB+VGOFJgAOl4SJoH8L84AAejgJ0ACYhiArhr2IxTLgfBsmyQhNUOTdC2Eww5eROc5Llw7YiOvN04MeOiAA8mAA4dI1nNQoCXKKRzAWl7OPLsxCXIQkqPBS3jUcAIGITBY0g6tImzV9SkrVEuMjYhkvknK0tEDK6uynK8sgQrcRKrNQxkGCbPXQLqsJYSPgqclOBkuoSXgCRZzhOIZCSIJ7EzGtswSOSEqUlT+QAAwAEgQHEmE01AICpIIuH2rZiHCDbQ3iZQrKUfTXFQCRWNk9iXlmuBZ2IJj4G8CoQMwOA1XUHVNPmphFvFStDOM4gOAkYAKg8CAwAqZw4GAZaRkrUrazDew4cWsMgkGqterrYR9iqQgJHR+79mAAAvDmKjahUux7HdYyEfZJmAURgDAXnUS3PsByFgAxKAsCl94JLgUyAFYaYEcQUsjMcSezFXaqhGFFudABOC2AHYAFoAAZTNt0yLeNhUlWxc3hCdK2LdtszbYAZhdGmrOq0tUU0tIICCOpUWRuAqL2YB7pbHUSUuTTIjSfxAeBzSZDhCpNJ1pghhk2OgA)

I checked the implementation of `CustomAttributeType` and it always takes the primitive type if available:

```ts
declare function CustomAttributeType<T>(
    base: T extends string ? 'string'
        : T extends number ? 'number'
        : T extends boolean ? 'boolean'
        : 'any'
): T extends string | number | boolean
    ? OpaquePrimitiveTypeName<T>
    : CustomAttributeTypeName<T>;
```

Is there a specific reason for this or could it be changed to handle union types?

**Comment by tywalch** [comment, marked as answer]
Hi @zirkelc 👋 

This is a necessity simply because of how TypeScript widens types. You can definitely roll your own option here to avoid the trappings of your union with `string`:

https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAUT5V2AT0IBlTADcsAYziEAwqlxUokgIJUq2AEaoqcACqbpcAHINJpAC+3Bg2YABEiHBGjrBO4QDc4kZQfFZgVAxOCKQAvBGaUKgYAPqZ2XAlfP5wieJUXqQACpi4qXyIAEpQOWD5ogA+RI5YfADmg0qiYDMZjQBcYADkKeoYDDFyS0mzYKuOG1Q6mfIAIgxui1bY4zuz++sxajDnl8M3Y0mBSaKr6d4YNp8DoIbo5TzePoRBh8TThMAMISoPgAaz4UBYfARQgsVhs9hGLjcEN8NQAPC1Ae0uj04AA+H6gMDuREooRqDQNYRiP4KNz+CBwSROVpQjosVTqLQACmmU12klgiEWijlu34nM0i3CmVwbPC+DVs30rSwqW1AEYDUaZsQMIYTNrdSiGNJpOEjYFDbsLoTXHgVTaMnB-IHdrsGt5tddRmNreHZuQ0FhyDBFo50EGvUHLJgwwm5lGIjHxvGC0nUCm4GmMhhMwnswnQjl8wnI3BFgCgSCwR5GlnPd7ZqN4AAPAPy8MBYWtVvh6QoucJrg4BA18ILssFvY2WS4dgdsAAbR1Ick4QAukGZo2C3ql+GV4h13qtwXgMA9uQLqQUjIoPubhgCicCaEIXByGA0hMBoDBIG08jAaBKDoLwuDXvcu4AQeiwnrmGCXhhgRZkON5yreyjlL0xEAJQ-KI-KSIKM6AgAdNIrjSqqMyMdqWRGGW+HavAXBls2h7hCWYwlPocHoB6gQ0axYxQNKdH1GezEimxHFUFxcq8dCTgCUOQkRCJYm0nODyHM8rwSQATAADA5ADMAC0FpOZ5ACsb42TExwwfZ2rOW57lORaEVWqRhYSQF1gEUOxGKcpqnqUAA

**Reply by zirkelc** [reply]
Great! Thanks for the quick reply!
